### PR TITLE
[no-merge] Check unused function params

### PR DESCRIPTION
**Pull request by som-snytt**
Both for explicit anon funs and those introduced
by for expr desugarings.

[PR status: improve warning message]

### CODE REVIEW

File: test/files/neg/warn-unused-params.check [hunk 1]
```diff
@@ -13,6 +13,12 @@ case class CaseyAtTheBat(k: Int)(s: String)        // warn
 warn-unused-params.scala:62: warning: parameter value readResolve in method f is never used
   def f(readResolve: Int) = 42           // warn
         ^
+warn-unused-params.scala:76: warning: parameter value i in value $anonfun is never used
+  val f = (i: Int) => 42
+            ^
+warn-unused-params.scala:78: warning: parameter value i in value $anonfun is never used
+  val g = for (i <- List(1)) yield 42
+               ^
```
Comment (hrhino) [inline test/files/neg/warn-unused-params.check]:
```scala
scala> val f = (_: Int) => 1
<console>:11: warning: parameter value x$1 in value $anonfun is never used
       val f = (_: Int) => 1
                 ^
```

I suggest that we shouldn't warn on unnamed (`WILDCARD`-named) params.

File: test/files/neg/warn-unused-params.check [hunk 2]
```diff
@@ -13,6 +13,12 @@ case class CaseyAtTheBat(k: Int)(s: String)        // warn
 warn-unused-params.scala:62: warning: parameter value readResolve in method f is never used
   def f(readResolve: Int) = 42           // warn
         ^
+warn-unused-params.scala:76: warning: parameter value i in value $anonfun is never used
```
Comment (hrhino) [inline test/files/neg/warn-unused-params.check:16]: ... "in this anonymous function is never used" if `p.owner.isAnonymousFunction`?

**Review by hrhino — COMMENTED** [review]
More pedantry is always... something.

Semi-relatedly, I turned on only `-Ywarn-unused:explicits` on my company's codebase yesterday; only a thousand lines needed changed. It's for the best.

File: test/files/neg/warn-unused-params.check (hunk 1 above)
Comment (som-snytt) [reply test/files/neg/warn-unused-params.check]: Good catch.

File: test/files/neg/warn-unused-params.check (hunk 2 above)
Comment (som-snytt) [reply test/files/neg/warn-unused-params.check:16]: I'll wait for final tweaking. The goal for the next push is to track usage during `for` desugaring or caramelization.

**Comment by som-snytt** [comment]
Forward ported, so this should be marked no-merge. Hopefully it will have a chance on 2.13.

File: test/files/neg/warn-unused-params.check (hunk 2 above)
Comment (som-snytt) [reply test/files/neg/warn-unused-params.check:16]: Done on linked PR.

**Review by adriaanm — APPROVED** [review]

**Comment by hrhino** [comment]
does this mean that #6924 is also good to go? It would be a real pity if this got lost somehow.

**Comment by xuwei-k** [comment]
regression?  🤔  https://github.com/scala/bug/issues/11175

**Pull request merged**